cpp/src/join/hash_join.cuh

```diff
@@ -137,6 +138,9 @@ size_type estimate_join_output_size(table_device_view build_table,
       h_size_estimate = size_estimate.value();
     }
 
+    // Detect overflow
+    CUDF_EXPECTS(h_size_estimate >= 0, "Maximum join output size exceeded");
```

Comment:
This is better than nothing, but it does not fix the underlying problem. An `int32_t` can overflow multiple times when two of them are multiplied together.  This means we can still have an overflow that produces a positive memory allocation.  The following patch on top of this one still hangs. As such I am happy to check this in as is so long as there is a follow on issue to fully fix the underlying problem.  I think if we do the size estimate as a `int64_t` it will solve the problem, but a `uint64_t` might be even better still, just in case cudf::size_type` goes to an `uint32_t`.

```diff
diff --git a/cpp/tests/join/join_tests.cpp b/cpp/tests/join/join_tests.cpp
index 9c0e2a9684..09c031d706 100644
--- a/cpp/tests/join/join_tests.cpp
+++ b/cpp/tests/join/join_tests.cpp
@@ -489,13 +489,13 @@ TEST_F(JoinTest, LeftJoinOnNulls)
 
 TEST_F(JoinTest, InnerJoinSizeOverflow)
 {
-  auto zero = cudf::make_numeric_scalar(cudf::data_type(cudf::type_id::INT32));
+  auto zero = cudf::make_numeric_scalar(cudf::data_type(cudf::type_id::INT8));
   zero->set_valid(true);
-  static_cast<cudf::scalar_type_t<int32_t> *>(zero.get())->set_value(0);
+  static_cast<cudf::scalar_type_t<int8_t> *>(zero.get())->set_value(0);
 
   // Should cause size overflow, raise exception
-  int32_t left  = 45000;
-  int32_t right = 450000;
+  int32_t left = 4;
+  int32_t right = 1073741825;
 
   auto col0_0 = cudf::make_column_from_scalar(*zero, left);
   auto col1_0 = cudf::make_column_from_scalar(*zero, right);

```

Reply:
Yeah, I just threw this in as a quick fix and as I mentioned I don't think it is sufficient in general. I don't like the idea of using unsigned types for sizes (many think making `size_t` unsigned was a mistake. I don't think we were ever discussing making `size_type` unsigned.

Reply:
OK, I've verified that your change hangs, and then fixed it by using 64-bit integer for the size estimate and properly detecting overflow.